codex-rs/login/src/server.rs

```diff
@@ -89,11 +118,35 @@ pub fn run_login_server(
     }
     let shutdown_flag = shutdown_flag.unwrap_or_else(|| Arc::new(AtomicBool::new(false)));
     let shutdown_flag_clone = shutdown_flag.clone();
+    let timeout_flag = Arc::new(AtomicBool::new(false));
+
+    // Channel used to signal completion to timeout watcher.
+    let (done_tx, done_rx) = mpsc::channel::<()>();
+
+    if let Some(timeout) = opts.login_timeout {
+        spawn_timeout_watcher(
+            done_rx,
+            timeout,
+            shutdown_flag.clone(),
+            timeout_flag.clone(),
+            server.clone(),
+        );
+    }
+
+    let server_for_thread = server.clone();
     let server_handle = thread::spawn(move || {
         while !shutdown_flag.load(Ordering::SeqCst) {
-            let req = match server.recv() {
+            let req = match server_for_thread.recv() {
                 Ok(r) => r,
-                Err(e) => return Err(io::Error::other(e)),
+                Err(e) => {
```

Comment:
Hmm, so in the case where `unblock()` is called by `ShutdownHandle::shutdown()`, it appears this will be an ordinary `std::io::Error` rather than something more specific:

https://github.com/tiny-http/tiny-http/blob/e2215636a73c7c3ac2d0a59d607454591f2cb2ac/src/lib.rs#L421

so checking `shutdown_flag` does seem to be the best wait to disambiguate the cause of the shutdown.